### DISCUSSION TITLE
Is it possible to use sync-wave for app?

### DISCUSSION THREAD

**Discussion by dmitry-mightydevops** [argoproj/argo-cd#6447] [Q&A]
Guys,

Is it possible to have a sync-wave control for apps using app of apps pattern?
https://argoproj.github.io/argo-cd/user-guide/sync-waves/

something like this, so we can have a sync order between the apps (i.e. app1, then app2 etc).

```
# helm show values grafana/grafana
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: ingress-nginx
  annotations:
    argocd.argoproj.io/sync-wave: "1"
  finalizers:
  - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    server: https://kubernetes.default.svc
    namespace: ingress-nginx
  project: default
  source:
    chart: ingress-nginx
    helm:
      values: |
        controller:
          config:
            use-http2: "true"
          nodeSelector:
            ops: "true"
          service:
            annotations:
              external-dns.alpha.kubernetes.io/hostname: domain.com
              service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
              service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
              service.beta.kubernetes.io/aws-load-balancer-type: nlb
        
          metrics:
            enabled: true
          podLabels: 
            app: "ingress-nginx"

    repoURL: https://kubernetes.github.io/ingress-nginx
    targetRevision: 3.33.0
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```

**Comment by jannfis** [comment]
Hey @dmitry-mightydevops, the short answer to this is, "it depends".

The long answer is, sync-waves apply to any given running sync operation. So, if you have an app-of-apps (let's call it `parent`) that spawns multiple child apps (let's call them `child1`, `child2` and `child3` here), and the child apps have a sync-wave annotation on the `Application` resource, the sync-waves apply _only_ during sync of `parent`. 

They will not apply upon sync of the child apps, so if all three child apps are out of sync, and auto-sync is enabled, they will just sync in any order as deemed right by the application controller.

So if your use-case is to sync those apps in a specific order at _bootstrap_ time (e.g. when they're initially created), this will work fine. You sync the `parent` app into the cluster, and the `child1` through `child3` will be created (and synced) according to their sync-waves. But as said, this will only happen when the `parent` app actually syncs. Furthermore, for Argo CD v1.8 and upwards, you will need to explicitly configure a [health check for Application resources](https://argo-cd.readthedocs.io/en/stable/operator-manual/upgrading/1.7-1.8/#health-assessement-of-argoprojioapplication-crd-has-been-removed) in order to make this work.
